src/components/button/button-base.tsx

```diff
@@ -0,0 +1,119 @@
+   * disabled; for example, to allow a tooltip to be displayed that explains why the button is disabled.
+   */
+  'aria-disabled'?: boolean | 'true' | 'false'
+  /** The button's label */
+  children?: ReactNode
+  /** Remove default padding. Only applies to tertiary buttons. */
+  hasNoPadding?: boolean
+  /** Icon to display on the left side */
+  iconLeft?: ReactNode
+  /** Icon to display on the right side */
+  iconRight?: ReactNode
+  /**
+   * Whether the button is in a busy state. A busy button will be `aria-disabled`, so will be focusable. However,
+   * click events will be ignored while it is busy.
+   */
+  isBusy?: boolean
+  /** Whether the button represents a destructive action */
+  isDestructive?: boolean
+  /** The size of the button */
+  size: 'small' | 'medium' | 'large'
+  /** Whether to use link-style appearance. Only applies to tertiary buttons. */
+  useLinkStyle?: boolean
+  /** The visual variant of the button */
+  variant: 'primary' | 'secondary' | 'tertiary'
+}
+
+export interface ButtonAsButtonProps extends CommonButtonBaseProps, ButtonHTMLAttributes<HTMLButtonElement> {
+  as: 'button'
+}
+
+export interface ButtonAsAnchorProps extends CommonButtonBaseProps, AnchorHTMLAttributes<HTMLAnchorElement> {
+  as: 'a'
+  href: string
+}
+
+export type ButtonBaseProps = ButtonAsButtonProps | ButtonAsAnchorProps
+
+/**
+ * A polymorphic button foundation that can render as either a button or anchor element.
+ * This component is used internally by the `Button` and `AnchorButton` components and
+ * should not be used directly by consumers.
+ */
+export function ButtonBase({
+  'aria-disabled': ariaDisabled,
+  as: Element,
+  className,
+  children,
+  hasNoPadding,
+  iconLeft,
+  iconRight,
+  isBusy,
+  isDestructive,
+  onClick,
+  size,
+  useLinkStyle,
+  variant,
+  ...rest
+}: ButtonBaseProps) {
+  // It's an icon-only button if there's no label text and only one icon
+  const isIconOnly = !children && (iconLeft || iconRight) && !(iconLeft && iconRight)
+
+  const handleClick = useCallback<MouseEventHandler<HTMLElement>>(
+    (event) => {
+      const element = event.currentTarget
+      // NOTE: Anchor elements CANNOT be disabled using the native `disabled` attribute, so we allow the
+      // `aria-disabled` attribute to disable them instead. Since click events will still be fired when
+      // `aria-disabled='true'`, we need to prevent any default action for the button from occuring, stop it
+      // propagating to ancestors and avoid calling the consumer-supplied `onClick` callback.
+      if (element.getAttribute('aria-disabled') === 'true') {
+        event.preventDefault()
+        event.stopPropagation()
+        return
+      }
+
+      // NOTE: We use a type assertion here to avoid having to narrow the type of `event` based on the specific
+      // `Element` type.
+      onClick?.(event as any)
```

Comment:
I think MouseEventHandler<HTMLElement> would be fine here - it's as generic as you can go, covers all bases!